enhancements/machine-config/custom-ignition-machineconfig.md

```diff
@@ -0,0 +1,199 @@
+---
+title: Store user ignition customizations in MachineConfig
+authors:
+  - "@hardys"
+  - "@celebdor"
+reviewers:
+  - "@celebdor"
+  - "@cgwalters"
+  - "@crawford"
+  - "@kirankt"
+  - "@runcom"
+  - "@stbenjam"
+  - "@yuqi-zhang"
+approvers:
+  - "@cgwalters"
+  - "@crawford"
+  - "@runcom"
+  - "@yuqi-zhang"
+
+creation-date: 2020-08-21
+last-updated: 2020-11-13
+status: implementable
+---
+
+# Store user ignition customizations in MachineConfig
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes changing the way installer ignition customizations
+are stored, so that instead of storing the modified pointer ignition in a
+Secret we include the user changes in a MachineConfig, such that the MCO
+can manage it, and it is included in the MCO rendered config.
+
+## Motivation
+
+The installer supports user [modification of the pointer ignition config](https://github.com/openshift/installer/blob/master/docs/user/customization.md#os-customization-unvalidated) it generates.  While this interface is marked as unvalidated, we know from [previous bug reports](https://bugzilla.redhat.com/show_bug.cgi?id=1881703) that some users are using it.
+
+This presents a problem for the plans to have the [MCO manage the pointer config](https://github.com/openshift/enhancements/blob/master/enhancements/machine-config/user-data-secret-managed.md)
+because it uses a static template that does not consider any user customizations, and for that reason the work was [reverted](https://github.com/openshift/machine-config-operator/pull/2126).
+
+Additionally, in some situations it is necessary to perform network configuration
+before it is possible to download the rendered config from the MCS in the ramdisk
+where ignition evaluates the pointer config.
+
+This leads to a chicken-egg problem, where a user could configure the network
+via ignition, but ignition cannot apply any config until it has downloaded
+any append/merge directives.
+
+We could solve that problem on some platforms (baremetal in particular) by
+just providing the fully rendered config to each host, bypassing the pointer
+config where config-drive size limits allow.  However this results in the [same
+issue with losing any user customizations](https://bugzilla.redhat.com/show_bug.cgi?id=1833483) provided directly via the pointer config.
+
+Discussions indicate we may want to deprecate/remove this ignition-config method of
```

Comment:
I don't see a way to do this without solving https://github.com/openshift/machine-config-operator/issues/1720
People definitely rely on customizing the pointer config with networking in UPI scenarios (as well as controlling the kernel cmdline).

Has the machine-specific problem been addressed in some other way in baremetal IPI?

Reply:
No we don't have a solution for the machine-specific problem in IPI - I'm hoping we can make progress on the MCO issue you mentioned to resolve that :+1: 

The aim of this enhancement is mostly to solve the bond+vlan case for IPI, where you don't need per-machine configs (like static ips), but you do need to avoid the pointer ignition download, because the vlan isn't configured yet.